eth/downloader/whitelist/service.go

```diff
@@ -3,133 +3,152 @@ package whitelist
-	}
-}
-
 var (
+	ErrMismatch = errors.New("mismatch error")
+	ErrNoRemote = errors.New("remote peer doesn't have a target block number")
+
 	ErrCheckpointMismatch = errors.New("checkpoint mismatch")
 	ErrLongFutureChain    = errors.New("received future chain of unacceptable length")
 	ErrNoRemoteCheckoint  = errors.New("remote peer doesn't have a checkoint")
 )
 
-// IsValidPeer checks if the chain we're about to receive from a peer is valid or not
-// in terms of reorgs. We won't reorg beyond the last bor checkpoint submitted to mainchain.
-func (w *Service) IsValidPeer(remoteHeader *types.Header, fetchHeadersByNumber func(number uint64, amount int, skip int, reverse bool) ([]*types.Header, []common.Hash, error)) (bool, error) {
-	// We want to validate the chain by comparing the last checkpointed block
-	// we're storing in `checkpointWhitelist` with the peer's block.
-	//
-	// Check for availaibility of the last checkpointed block.
-	// This can be also be empty if our heimdall is not responding
-	// or we're running without it.
-	if len(w.checkpointWhitelist) == 0 {
-		// worst case, we don't have the checkpoints in memory
-		return true, nil
-	}
+type Service struct {
+	checkpointService
+	milestoneService
+}
 
-	// Fetch the last checkpoint entry
-	lastCheckpointBlockNum := w.checkpointOrder[len(w.checkpointOrder)-1]
-	lastCheckpointBlockHash := w.checkpointWhitelist[lastCheckpointBlockNum]
+func NewService(db ethdb.Database) *Service {
+	var checkpointDoExist = true
+
+	checkpointNumber, checkpointHash, err := rawdb.ReadFinality[*rawdb.Checkpoint](db)
 
-	// todo: we can extract this as an interface and mock as well or just test IsValidChain in isolation from downloader passing fake fetchHeadersByNumber functions
-	headers, hashes, err := fetchHeadersByNumber(lastCheckpointBlockNum, 1, 0, false)
 	if err != nil {
-		return false, fmt.Errorf("%w: last checkpoint %d, err %v", ErrNoRemoteCheckoint, lastCheckpointBlockNum, err)
+		checkpointDoExist = false
 	}
 
-	if len(headers) == 0 {
-		return false, fmt.Errorf("%w: last checkpoint %d", ErrNoRemoteCheckoint, lastCheckpointBlockNum)
+	var milestoneDoExist = true
+
+	milestoneNumber, milestoneHash, err := rawdb.ReadFinality[*rawdb.Milestone](db)
+	if err != nil {
+		milestoneDoExist = false
 	}
 
-	reqBlockNum := headers[0].Number.Uint64()
-	reqBlockHash := hashes[0]
+	locked, lockedSprintNumber, lockedSprintHash, lockedMilestoneIDs, err := rawdb.ReadLockField(db)
+	if err != nil || !locked {
+		locked = false
+		lockedMilestoneIDs = make(map[string]struct{})
+	}
 
-	// Check against the checkpointed blocks
-	if reqBlockNum == lastCheckpointBlockNum && reqBlockHash == lastCheckpointBlockHash {
-		return true, nil
+	order, list, err := rawdb.ReadFutureMilestoneList(db)
+	if err != nil {
+		order = make([]uint64, 0)
+		list = make(map[uint64]common.Hash)
 	}
 
-	return false, ErrCheckpointMismatch
+	return &Service{
+		//fixme: try to restore from DB
```

Comment:
Is this for future optimisations?

Reply:
Forgot to remove the comments....Actually restoring it from Db